kernel: Add _STATIC variant for object definitions

hubpav opened the pull request:
All existing DEFINE_* macros define objects in global scope. However,
if the same object name is used across multiple module, the result
is a conflict during linkage phase. The goal is to provide _STATIC
to these DEFINE_* macros, so the user can safely use them in a private
scope.

This commit is fully backward compatible - i.e. it does not change the
implementation of the existing DEFINE_* macros and only adds new ones.

Also, this commit fixes macro formatting inconsistencies across
the whole module. It was inconvenient to create such a vast commit
without fixing the style in the first place.

Signed-off-by: Pavel Hübner <pavel.hubner@hardwario.com>

hubpav commented on include/kernel.h:
```diff
@@ -4401,7 +4401,7 @@ struct k_msgq_attrs {
 	static char __noinit __aligned(q_align)				\
 		_k_fifo_buf_##q_name[(q_max_msgs) * (q_msg_size)];	\
-	STRUCT_SECTION_ITERABLE(k_msgq, q_name) =			\
+	static STRUCT_SECTION_ITERABLE(k_msgq, q_name) =		\
```
@carlescufi Yep, that's right - `K_PIPE_DEFINE` is also affected. I have just walked through `kernel.h` and all occurrences of `STRUCT_SECTION_ITERABLE` and we should actually introduce `#define <...>_DEFINE_STATIC` suffixes to all helper macros, where there is no solution to just prepend `static` in front of those (i.e. they define some auxiliary buffer before the actual object).

I have made a similar PR before (already merged #39683):

* `K_MEM_SLAB_DEFINE` - Statically define and initialize a memory slab in a public (non-static) scope.
* `K_MEM_SLAB_DEFINE_STATIC` - Statically define and initialize a memory slab in a private (static) scope.

So, this original commit is flawed in a way it prevents to `extern` such definition elsewhere. Although not the best way, still the API could keep this option available.

Based on this logic, we can discuss the availability of:

* `K_MSGQ_DEFINE`
* `K_MSGQ_DEFINE_STATIC`
* `K_PIPE_DEFINE`
* `K_PIPE_DEFINE_STATIC`
* etc.

But for `K_MUTEX_DEFINE`, it is not necessary, as `static K_MUTEX_DEFINE` does the job. The other perspective: For the sake of API clarity, all `_DEFINE` macros should have their `_DEFINE_STATIC` counterparts.

What do you think? Thanks.

carlescufi replied:
Thanks @hubpav, I will let @andyross, @dcpleung and @peter-mitsis weigh in here.

peter-mitsis replied:
As you correctly point out, the K_MSGQ_DEFINE() macro defines two objects--a buffer and the k_msgq instance. It is logical for the buffer portion to be marked as 'static' (or private to the file). However, the k_msgq instance may be referenced from another file, so it is not marked as such.

I think that you are also correct in pointing out the need for a static (private to the file) instance of the k_msgq object. Since the instance created by K_MSGQ_DEFINE() is meant to in theory be available to other modules, I think your idea of creating a new macro K_MSGQ_DEFINE_STATIC() is the best route forward--particularly since there is precendent with K_MEM_SLAB_DEFINE_STATIC().

andyross reviewed (changes requested):
Am I misunderstanding?  This looks like an API breakage to me.  Right now if you have a `K_MSGQ_DEFINE(my_queue, ...)` in one C file, you can using it anywhere else in your app as "`extern struct k_msgq *my_queue;`"  (to be fair: it would be nicer if we had a `K_MSGQ_DECLARE()` to do that).   This breaks that kind of usage for existing code that was doing things correctly.

I suspect the real bug here is that with other data structures, it's often possible to put a "static" in front of the macro to define a local object.  (Linux does this a lot, for example).  And that's not possible here because of the need for the prefixed buffer definition.

hubpav commented:
> Am I misunderstanding? This looks like an API breakage to me. Right now if you have a `K_MSGQ_DEFINE(my_queue, ...)` in one C file, you can use it anywhere else in your app as "`extern struct k_msgq *my_queue;`" (to be fair: it would be nicer if we had a `K_MSGQ_DECLARE()` to do that). This breaks that kind of usage for existing code that was doing things correctly.
> 
> I suspect the real bug here is that with other data structures, it's often possible to put a "static" in front of the macro to define a local object. (Linux does this a lot, for example). And that's not possible here because of the need for the prefixed buffer definition.

@andyross Yes, correct - in the end, this could be a breaking change and I will restructure the work.

However, in the comment above, I have proposed a change (and @peter-mitsis has sort of supported it) that we should keep the current definition of `K_MSGQ_DEFINE` as-is, but introduce new definitions for purely private scoping:

- `K_PIPE_DEFINE_STATIC`
- `K_MSGQ_DEFINE_STATIC`

And maybe some others - I have not yet thoroughly traversed the whole `kernel.h` but I will refine my commit and we can do another round...

henrikbrixandersen commented:
@hubpav Do you plan on picking this up again?

hubpav commented:
> @hubpav Do you plan on picking this up again?

Yep, absolutely - it is on my mind twice a week :) I will be on it this month definitely, thanks.

henrikbrixandersen commented:
> Yep, absolutely - it is on my mind twice a week :) I will be on it this month definitely, thanks.

Time flies - a month has passed :)

hubpav commented:
> > Yep, absolutely - it is on my mind twice a week :) I will be on it this month definitely, thanks.
> 
> Time flies - a month has passed :)

I love the reminder :) Thanks a lot @henrikbrixandersen!

In the end, it was a bit more effort than estimated and it resulted in a fair number of code lines changed but I think it is the right direction.

Actually, it should be a preferred option to guide developers to prefer e.g. `K_MUTEX_DEFINE_STATIC(...)` instead of `static K_MUTEX_DEFINE(...)`. The reason for that is that it removes the ambiguity between such primitives and more complex types which span across multiple definitions.

henrikbrixandersen commented:
@hubpav Please rebase on `main` and resolve the conflicts.

hubpav commented:
@henrikbrixandersen Done

dcpleung reviewed (commented):
There are quite a few whitespace changes whic are not really related to adding _STATIC definitions. Please separate them into separate commit so it is easier to review.

hubpav committed:
kernel: Add _STATIC variant for object definitions

All existing DEFINE_* macros define objects in global scope. However,
if the same object name is used across multiple module, the result
is a conflict during linkage phase. The goal is to provide _STATIC
to these DEFINE_* macros, so the user can safely use them in a private
scope.

This commit is fully backward compatible - i.e. it does not change the
implementation of the existing DEFINE_* macros and only adds new ones.

Signed-off-by: Pavel Hübner <pavel.hubner@hardwario.com>

hubpav committed:
kernel: Unify backslash indent in multi-line macro

Across the whole kernel.h, multiple backslash macro continuation
styles have been used. That made the implementation inconsistent
and hard to read.

Also, the space and tab indentation has been fixed
to allow various indentation preferences in user's editor.

Signed-off-by: Pavel Hübner <pavel.hubner@hardwario.com>

ceolin commented:
I think we should have, at least, one test using these macro variants. Also I don't know if copy&paste the documentation is the best approach. Maybe reference the original macro and just document the difference would be better, but I don't what is the guideline for it.

hubpav commented:
@ceolin I will be happy to write the test but I will need a bit of guidance here. Shall I prepare the test that will define all static initializers with equal names in two files (modules)? The test would succeed if the application builds, so no ztest will be involved. Or do you have any other ideas about this?

About the documentation: We have touched this point in the very first PR of mine which was addressing the same thing - adding `K_MEM_SLAB_DEFINE_STATIC `:

https://github.com/zephyrproject-rtos/zephyr/pull/39683#discussion_r735660227

Basically, I have no issue to change that though. Will give it a few days if someone else raises their hand...

carlescufi commented:
> @ceolin I will be happy to write the test but I will need a bit of guidance here. Shall I prepare the test that will define all static initializers with equal names in two files (modules)? The test would succeed if the application builds, so no ztest will be involved. Or do you have any other ideas about this?
> 
> About the documentation: We have touched this point in the very first PR of mine which was addressing the same thing - adding `K_MEM_SLAB_DEFINE_STATIC `:
> 
> [#39683 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/39683#discussion_r735660227)
> 
> Basically, I have no issue to change that though. Will give it a few days if someone else raises their hand...

@dcpleung and @andyross guidance needed here.

andyross commented:
Is the question just about tests?  In this case these macros are all very simple, I think a straightforward coverage case that did nothing but define an object using them and use it once would be more than sufficient.  We just want to know it works.

hubpav commented:
> Is the question just about tests? In this case these macros are all very simple, I think a straightforward coverage case that did nothing but define an object using them and use it once would be more than sufficient. We just want to know it works.

@andyross This sounds like a reasonable approach and I agree that automated tests are not truly needed when this gets verified once.

So, I will do the following:

1. Rebase the code, of course :)
2. Perform a manual coverage tests on all the macros.
3. Update the documentation, so we won't duplicate text just for the sake of private/public scoping.

Anything else that should be done?

I can do it no earlier than this weekend.

andyross commented:
Oh, to be clear: I meant check in a ztest test case.  I'm just saying it doesn't have to be more than a few lines of code.  Create a semaphore, take/give, etc...

henrikbrixandersen commented:
This needs rebasing.

hubpav commented:
> This needs rebasing.

Done.
